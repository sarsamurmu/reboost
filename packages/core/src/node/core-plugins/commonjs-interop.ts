import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';

import { ReboostPlugin } from '../index';

export const CommonJSInteropPlugin: ReboostPlugin = {
  name: 'core-commonjs-interop-plugin',
  transformAST(ast, { traverse }, filePath) {
    let program: NodePath<t.Program>;
    let importerIdentifier: t.Identifier;
    const replacements: [NodePath<t.ImportDeclaration>, t.ImportDeclaration, t.VariableDeclarator[]][] = [];

    traverse(ast, {
      Program(path) {
        program = path;
      },
      ImportDeclaration(path) {
        const declarators: t.VariableDeclarator[] = [];
        const identifier = path.scope.generateUidIdentifier('$import');
        if (!importerIdentifier) importerIdentifier = path.scope.generateUidIdentifier('$importer');

        path.node.specifiers.forEach((specifier) => {
          let usage;
          let importedName;
          const localName = specifier.local.name;
          const commons = [
            t.callExpression(
              t.identifier('__reboost_resolve'),
              [t.stringLiteral(path.node.source.value)]
            ),
            t.stringLiteral(filePath)
          ];

          if (t.isImportDefaultSpecifier(specifier)) {
            usage = 'Default';
          } else if (t.isImportNamespaceSpecifier(specifier)) {
            usage = 'All';
          } else if (t.isImportSpecifier(specifier)) {
            usage = 'Member';
            importedName = specifier.imported.name;
          }

          declarators.push(
            t.variableDeclarator(
              t.identifier(localName),
              t.callExpression(
                t.memberExpression(
                  importerIdentifier,
                  t.identifier(usage)
                ),
                importedName ? [
                  identifier,
                  t.stringLiteral(importedName),
                  ...commons
                ] : [identifier, ...commons]
              )
            )
          );
        });

        replacements.push([
          path,
          t.importDeclaration([
            t.importNamespaceSpecifier(identifier)
          ], t.stringLiteral(path.node.source.value)),
          declarators
        ]);
      }
    });

    if (importerIdentifier) {
      replacements.forEach(([path, replacement, declarators]) => {
        path.replaceWith(replacement);
        if (declarators.length) {
          const constDeclaration = t.variableDeclaration('const', declarators);
          path.insertAfter(constDeclaration);
        }
      });

      program.node.body.unshift(
        t.importDeclaration([
          t.importDefaultSpecifier(importerIdentifier)
        ], t.stringLiteral(`#/importer`))
      );
    }
  }
}
