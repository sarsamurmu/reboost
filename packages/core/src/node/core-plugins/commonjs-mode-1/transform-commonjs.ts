import * as t from '@babel/types';
import traverse, { NodePath } from '@babel/traverse';

import { builtinModules } from 'module';

const getReboostResolveCall = (source: string) => t.callExpression(
  t.identifier('__reboost_resolve'),
  [t.stringLiteral(source)]
);

export const transformCommonJS = (ast: t.Node, filePath: string, id: string) => {
  let program: NodePath<t.Program>;
  let usedModuleExports: boolean;
  let usedExports: boolean;
  let modImports: t.ImportDeclaration[];
  let importIdentifierMap: Record<string, t.Identifier>;
  const importerIdentifier: t.Identifier = t.identifier(`importer_${id}`);
  let usedImporter: boolean;
  let replacements: [NodePath<t.ImportDeclaration>, t.ImportDeclaration, t.VariableDeclarator[]][];
  let importIdx = 0;

  traverse(ast, {
    Program(path) {
      program = path;
      if (path.scope.hasGlobal('module')) usedModuleExports = true;
      if (path.scope.hasGlobal('exports')) usedExports = true;
    },
    ImportDeclaration(path) {
      if (path.node.specifiers.length === 0) return;

      const declarators: t.VariableDeclarator[] = [];
      const identifier = t.identifier(`import_${importIdx++}_${id}`);

      usedImporter = true;

      path.node.specifiers.forEach((specifier) => {
        let usage;
        let importedName;
        const localName = specifier.local.name;
        const commons = [
          getReboostResolveCall(path.node.source.value),
          t.stringLiteral(filePath)
        ];

        if (t.isImportDefaultSpecifier(specifier)) {
          usage = 'Default';
        } else if (t.isImportNamespaceSpecifier(specifier)) {
          usage = 'All';
        } else if (t.isImportSpecifier(specifier)) {
          usage = 'Member';
          importedName = t.isIdentifier(specifier.imported)
            ? specifier.imported.name
            : specifier.imported.value;
        }

        declarators.push(
          t.variableDeclarator(
            t.identifier(localName),
            t.callExpression(
              t.memberExpression(
                importerIdentifier,
                t.identifier(usage)
              ),
              importedName ? [
                identifier,
                t.stringLiteral(importedName),
                ...commons
              ] : [identifier, ...commons]
            )
          )
        );
      });

      (replacements || (replacements = [])).push([
        path,
        t.importDeclaration([
          t.importNamespaceSpecifier(identifier)
        ], t.stringLiteral(path.node.source.value)),
        declarators
      ]);
    },
    CallExpression(path) {
      if (
        t.isIdentifier(path.node.callee, { name: 'require' }) &&
        path.node.arguments.length === 1 &&
        t.isStringLiteral(path.node.arguments[0]) &&
        !path.scope.hasBinding('require')
      ) {
        const importPath = path.node.arguments[0].value;
        const importIdentifier = (importIdentifierMap || (importIdentifierMap = {}))[importPath] ||
          t.identifier(`imported_${importIdx++}_${id}`);

        // Don't resolve built-in modules like path, fs, etc.
        if (builtinModules.includes(importPath)) return;

        if (!(importPath in importIdentifierMap)) {
          importIdentifierMap[importPath] = importIdentifier;
          (modImports || (modImports = [])).push(
            t.importDeclaration(
              [t.importNamespaceSpecifier(importIdentifier)],
              t.stringLiteral(importPath)
            )
          );
        }

        usedImporter = true;

        path.replaceWith(
          t.callExpression(
            t.memberExpression(
              importerIdentifier,
              t.identifier('All'),
            ),
            [
              importIdentifier,
              getReboostResolveCall(importPath),
              t.stringLiteral(filePath)
            ]
          )
        );
      }
    },
  });

  if (usedModuleExports || usedExports) {
    if (usedModuleExports) {
      program.node.body.unshift(
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('module'),
              t.objectExpression([
                t.objectProperty(
                  t.identifier('exports'),
                  usedExports ? t.identifier('exports') : t.objectExpression([]),
                  false,
                  usedExports
                )
              ])
            ),
          ]
        )
      );
    }

    if (usedExports) {
      program.node.body.unshift(
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('exports'),
              t.objectExpression([])
            )
          ]
        )
      );
    }

    program.node.body.push(
      t.exportNamedDeclaration(
        t.variableDeclaration(
          'var',
          [
            t.variableDeclarator(
              t.identifier('__cjsExports'),
              usedModuleExports
                ? t.memberExpression(t.identifier('module'), t.identifier('exports'))
                : t.identifier('exports')
            )
          ]
        )
      )
    );
  }

  if (modImports) program.node.body.unshift(...modImports);

  if (usedImporter) {
    if (replacements) {
      replacements.forEach(([path, replacement, declarators]) => {
        if (declarators.length) {
          path.insertAfter(t.variableDeclaration('const', declarators));
        }
        path.replaceWith(replacement);
      });
    }

    program.node.body.unshift(
      t.importDeclaration([
        t.importDefaultSpecifier(importerIdentifier)
      ], t.stringLiteral('#/importer'))
    );
  }
}
