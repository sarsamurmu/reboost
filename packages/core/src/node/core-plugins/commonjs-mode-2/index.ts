import anymatch from 'anymatch';

import { ReboostInstance, ReboostPlugin } from '../../index';
import { transformCommonJSToES6 } from './commonjs-to-es6';
import { uniqueID } from '../../utils';

export { transformCommonJSToES6 as runTransformation }

export const CommonJSMode2Plugin = (instance: ReboostInstance): ReboostPlugin => {
  const test = (file: string) => (
    !anymatch(instance.config.commonJSInterop.exclude, file) &&
    anymatch(instance.config.commonJSInterop.include, file)
  );

  return {
    name: 'core-commonjs-mode-2-plugin',
    getCacheKey: ({ serializeObject }) => serializeObject({
      include: instance.config.commonJSInterop.include,
      exclude: instance.config.commonJSInterop.exclude
    }),
    transformAST(ast, _, filePath) {
      if (test(filePath)) {
        transformCommonJSToES6(ast, uniqueID(6));
      }
    }
  }
}
