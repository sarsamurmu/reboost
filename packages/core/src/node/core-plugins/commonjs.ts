import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';

import { builtinModules } from 'module';

import { ReboostPlugin } from '../index';
import { uniqueID } from '../utils';

export const isDeclared = (path: NodePath<any>, identifierName: string, findInProgram = true) => {
  const findInDeclarators = (id: t.LVal) => {
    let found = false;
    if (t.isIdentifier(id, { name: identifierName })) {
      found = true;
    } else if (t.isArrayPattern(id)) {
      id.elements.forEach((element) => {
        if (found) return;
        found = findInDeclarators(element);
      });
    } else if (t.isObjectPattern(id)) {
      id.properties.forEach((property) => {
        if (found) return;

        if (t.isRestElement(property)) {
          found = t.isIdentifier(property.argument, { name: identifierName });
        } else {
          found = findInDeclarators(property.value as any);
        }
      });
    }
    return found;
  }

  const findInParent = (astPath: NodePath<any>): boolean => {
    let found = false;

    const parent: NodePath<t.BlockStatement | t.Program> = astPath.findParent(
      (p) => p.isBlockStatement() || (findInProgram && p.isProgram())
    ) as any;
    if (parent) {
      parent.node.body.forEach((item) => {
        if (found) return;

        if (t.isVariableDeclaration(item)) {
          item.declarations.forEach((declaration) => {
            if (found) return;

            found = findInDeclarators(declaration.id);
          });
        }
      });

      if (!found) return findInParent(parent);
    }

    return found;
  }

  return findInParent(path);
}

export const CommonJSPlugin: ReboostPlugin = {
  name: 'core-commonjs-plugin',
  transformAST(ast, { traverse }) {
    let program: NodePath<t.Program>;
    let cjsModule = false;
    const uid = uniqueID(4);
    const modImports: t.ImportDeclaration[] = [];
    const importIdentifierMap = {} as Record<string, t.Identifier>;

    traverse(ast, {
      Program(path) {
        program = path;
      },
      CallExpression(path) {
        if (
          t.isIdentifier(path.node.callee, { name: 'require' }) &&
          t.isStringLiteral(path.node.arguments[0])
        ) {
          const importPath = path.node.arguments[0].value;
          const importIdentifier = importIdentifierMap[importPath] || path.scope.generateUidIdentifier(`$imported_${uid}`);

          // Don't resolve built-in modules like path, fs, etc.
          if (builtinModules.includes(importPath)) return;

          if (!(importPath in importIdentifierMap)) {
            importIdentifierMap[importPath] = importIdentifier;
            modImports.push(
              t.importDeclaration(
                [t.importNamespaceSpecifier(importIdentifier)],
                t.stringLiteral(importPath)
              )
            );
          }

          path.replaceWith(importIdentifier);
        }
      },
      MemberExpression(path) {
        if (cjsModule) return;
        if (
          t.isIdentifier(path.node.object, { name: 'module' }) &&
          t.isIdentifier(path.node.property, { name: 'exports' })
        ) {
          if (!isDeclared(path, 'module')) cjsModule = true;
        } else if (t.isIdentifier(path.node.object, { name: 'exports' })) {
          if (!isDeclared(path, 'exports')) cjsModule = true;
        }
      }
    });

    if (cjsModule) {
      const moduleExportsExp = t.memberExpression(t.identifier('module'), t.identifier('exports'));

      program.node.body.unshift(
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('module'),
              t.objectExpression([
                t.objectProperty(
                  t.identifier('exports'),
                  t.objectExpression([])
                )
              ])
            ),
          ]
        ),
        t.variableDeclaration(
          'const',
          [
            t.variableDeclarator(
              t.identifier('exports'),
              moduleExportsExp
            )
          ]
        )
      );

      program.node.body.push(
        t.exportNamedDeclaration(
          t.variableDeclaration(
            'const',
            [t.variableDeclarator(
              t.identifier('__cjsExports'),
              moduleExportsExp
            )]
          )
        )
      );
    }

    program.node.body.unshift(...modImports);
  }
}
