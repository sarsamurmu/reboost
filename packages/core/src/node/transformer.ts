import { parse } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import generate, { GeneratorOptions } from '@babel/generator';
import * as babelTypes from '@babel/types';
import { codeFrameColumns, SourceLocation } from '@babel/code-frame';
import anymatch from 'anymatch';
import { RawSourceMap, SourceMapConsumer } from 'source-map';
import chalk from 'chalk';
import MagicString from 'magic-string';

import fs from 'fs';
import path from 'path';

import { ReboostPlugin, PluginContext } from './index';
import { getConfig, getAddress } from './shared';
import { mergeSourceMaps, bind } from './utils';
import { resolveModule } from './core-plugins/resolver';

const fixPath = (pathString: string) => pathString.replace(/\\/g, '/');

const getCompatibleSourceMap = (map: RawSourceMap) => {
  const config = getConfig();

  map.sourceRoot = 'reboost:///';

  map.sources = map.sources.map((sourcePath: string) => {
    if (path.isAbsolute(sourcePath)) return fixPath(path.relative(config.rootDir, sourcePath));
    return fixPath(sourcePath);
  });

  map.sourcesContent = [];
  map.sources.forEach((sourcePath) => {
    const absolutePath = path.join(config.rootDir, sourcePath);
    if (fs.existsSync(absolutePath)) {
      map.sourcesContent.push(fs.readFileSync(absolutePath).toString());
    } else {
      console.log(chalk.red(`Unable to find file "${absolutePath}". Required for source map generation.`));
      map.sourcesContent.push(`Unable to find file in "${absolutePath}".`);
    }
  });

  map.file = undefined;

  return map;
}

const getPluginContext = (filePath: string, mergedDependencies: string[]): PluginContext => ({
  address: getAddress(),
  config: getConfig(),
  addDependency(dependency) {
    mergedDependencies.push(dependency);
  },
  chalk,
  getCompatibleSourceMap,
  MagicString,
  mergeSourceMaps,
  resolveModule
})

const handleError = ({ message }: Error, dependencies: string[]) => {
  console.log(chalk.red(message));

  return {
    code: `console.error(${JSON.stringify('[reboost] ' + message)})`,
    dependencies,
    error: true
  }
}

let pluginsInitiated = false;
let resolveHooks: ReboostPlugin['resolve'][];
let loadHooks: ReboostPlugin['load'][];
let transformContentHooks: ReboostPlugin['transformContent'][];
let transformIntoJSHooks: ReboostPlugin['transformIntoJS'][];
let transformASTHooks: ReboostPlugin['transformAST'][];

export const transformFile = async (filePath: string): Promise<{
  code: string;
  dependencies: string[];
  map?: string;
  imports?: string[];
  error?: boolean;
}> => {
  if (!pluginsInitiated) {
    const def = (a: any) => !!a;
    const plugins = getConfig().plugins.filter(def);

    resolveHooks = plugins.map((plugin) => plugin.resolve).filter(def);
    loadHooks = plugins.map((plugin) => plugin.load).filter(def);
    transformContentHooks = plugins.map((plugin) => plugin.transformContent).filter(def);
    transformIntoJSHooks = plugins.map((plugin) => plugin.transformIntoJS).filter(def);
    transformASTHooks = plugins.map((plugin) => plugin.transformAST).filter(def);
    
    pluginsInitiated = true;
  }

  let code: string;
  let sourceMap: RawSourceMap;
  let ast: babelTypes.Node;
  let type: string;
  let errorOccurred = false;
  const imports: string[] = [];
  const dependencies: string[] = [];
  const pluginContext = getPluginContext(filePath, dependencies);

  for (const hook of loadHooks) {
    let result = await bind(hook, pluginContext)(filePath);
    if (result) {
      ({ code } = result);
      ({ type } = result);
      if (result.map) sourceMap = result.map;
      break;
    }
  }

  for (const hook of transformContentHooks) {
    const result = await bind(hook, pluginContext)({ code, type, map: sourceMap }, filePath);
    if (result) {
      if (result instanceof Error) return handleError(result, dependencies);

      ({ code } = result);
      if (result.map) {
        // Here source maps sources can be null, like when source map is generated using MagicString (npm package)
        result.map.sources = result.map.sources.map((sourcePath) => !sourcePath ? filePath : sourcePath);
        sourceMap = sourceMap ? await mergeSourceMaps(sourceMap, result.map) : result.map;
      }
      if (result.type) ({ type } = result);
    }
  }

  for (const hook of transformIntoJSHooks) {
    const result = await bind(hook, pluginContext)({
      code,
      type,
      map: sourceMap
    }, filePath);

    if (result) {
      if (result instanceof Error) return handleError(result, dependencies);

      ({ code } = result);
      sourceMap = result.inputMap;
      type = 'js';
      break;
    }
  }

  if (type !== 'js') {
    console.log(chalk.red(`File with type "${type}" is not supported. You may need proper loader to transform this kind of files into JS.`));
  }

  try {
    ast = parse(code, {
      sourceType: 'module'
    });
  } catch (e) {
    let message = '';
    let consoleMessage = '';
    let frameMessage = e.message.replace(/\s*\(.*\)$/, '');
    let rawCode = code;
    let location: SourceLocation['start'] = e.loc;
    let unableToLocateFile = false;
    message += `Error while parsing "${filePath}"\n`;
    message += 'You may need proper loader to handle this kind of files.\n\n';

    consoleMessage += chalk.red(message);

    if (sourceMap) {
      const consumer = await new SourceMapConsumer(sourceMap);
      const originalLoc = consumer.originalPositionFor(e.loc);
      if (originalLoc.source) {
        const originalCode = consumer.sourceContentFor(originalLoc.source);
        if (originalCode) {
          rawCode = originalCode;
          location = originalLoc;
          location.column = location.column || 1;
        } else {
          const absPathToSource = path.join(getConfig().rootDir, originalLoc.source);
          if (fs.existsSync(absPathToSource)) {
            rawCode = fs.readFileSync(absPathToSource).toString();
            location = originalLoc;
            location.column = location.column || 1;
          } else {
            unableToLocateFile = true;
          }
        }
      }
    }

    if (unableToLocateFile) {
      let unableToLocateMsg = 'We are unable to locate the original file. ';
      unableToLocateMsg += 'This is not accurate, but it may help you at some point.\n\n';
      
      message += unableToLocateMsg + codeFrameColumns(code, { start: e.loc }, {
        message: frameMessage
      });

      consoleMessage += unableToLocateMsg + codeFrameColumns(code, { start: e.loc }, {
        highlightCode: true,
        message: frameMessage
      });
    } else {
      message += codeFrameColumns(rawCode, { start: location }, {
        message: frameMessage
      });

      consoleMessage += codeFrameColumns(rawCode, { start: location }, {
        highlightCode: true,
        message: frameMessage
      });
    }

    console.log(consoleMessage);

    return {
      code: `console.error('[reboost] ' + ${JSON.stringify(message)});`,
      dependencies,
      error: true
    }
  }

  for (const hook of transformASTHooks) await bind(hook, pluginContext)(ast, { traverse, types: babelTypes }, filePath);

  const resolvedPathMap = {} as Record<string, any>;
  const resolvePath = async (source: string, filePath: string) => {
    const keyStr = `${filePath}||${source}`;
    if (resolvedPathMap[keyStr]) return resolvedPathMap[keyStr];
    for (const hook of resolveHooks) {
      const resolvedPath = await hook(source, filePath);
      if (resolvedPath) {
        resolvedPathMap[keyStr] = resolvedPath;
        return resolvedPath
      }
    }
    console.log(chalk.red(`[reboost] Unable to resolve path "${source}" of "${filePath}"`));
    return null;
  }

  const resolveDeps = async (astPath: NodePath<babelTypes.ImportDeclaration> | NodePath<babelTypes.ExportDeclaration>) => {
    if ((astPath.node as any).source) {
      const source: string = (astPath.node as any).source.value;
      
      if (source === 'reboost/hmr') {
        (astPath.node as any).source.value = `/hmr?q=${encodeURI(filePath)}`;
      } else {
        let finalPath = null;
        let routed = false;
        if (source.startsWith('#/')) {
          finalPath = source.replace(/^#/, '');
          routed = true;
        } else {
          const resolvedPath = await resolvePath(source, filePath);
          if (resolvedPath) {
            if (resolvedPath.startsWith('#/')) {
              finalPath = resolvedPath.replace(/^#/, '');
              routed = true;
            } else {
              finalPath = resolvedPath;
              imports.push(resolvedPath);
            }
          } else {
            errorOccurred = true;
          }
        }

        (astPath.node as any).source.value = getAddress() + (routed
          ? encodeURI(finalPath)
          : finalPath
            ? `/transformed?q=${encodeURI(finalPath)}`
            : `/unresolved?import=${encodeURI(source)}`);
      }
    }
  }

  const promiseFunctions: (() => Promise<void>)[] = [];
  let astProgram: NodePath<babelTypes.Program>;

  traverse(ast, {
    Program(astPath) {
      astProgram = astPath;
    },
    ImportDeclaration(astPath) {
      promiseFunctions.push(async () => {
        await resolveDeps(astPath);
      });
      return false;
    },
    ExportDeclaration(astPath) {
      promiseFunctions.push(async () => {
        await resolveDeps(astPath);
      });
      return false;
    },
    CallExpression(astPath) {
      const t = babelTypes;
      if (t.isIdentifier(astPath.node.callee, { name: '__reboost_resolve' })) {
        promiseFunctions.push(async () => {
          astPath.replaceWith(
            t.stringLiteral(
              await resolvePath((astPath.node.arguments[0] as babelTypes.StringLiteral).value, filePath)
            )
          );
        });
      } else if (t.isImport(astPath.node.callee)) {
        // Rewrite dynamic imports
        const importDeclarations = astProgram.get('body').filter((p) => p.isImportDeclaration());
        let importerDeclaration = importDeclarations.find(
          (dec) => (dec as NodePath<babelTypes.ImportDeclaration>).node.source.value.includes('#/importer')
        ).node as babelTypes.ImportDeclaration;
        if (!importerDeclaration) {
          const identifier = astPath.scope.generateUidIdentifier('$importer');
          importerDeclaration = t.importDeclaration([
            t.importDefaultSpecifier(identifier)
          ], t.stringLiteral(`#/importer`));
          astProgram.node.body.unshift(importerDeclaration);
        }
        const importerIdentifierName = importerDeclaration.specifiers[0].local.name;
        astPath.replaceWith(
          t.callExpression(
            t.memberExpression(
              t.identifier(importerIdentifierName),
              t.identifier('Dynamic')
            ),
            [
              t.stringLiteral((astPath.node.arguments[0] as babelTypes.StringLiteral).value),
              t.stringLiteral(filePath)
            ]
          )
        );
      }
    }
  });

  for (const execute of promiseFunctions) await execute();

  const sourceMapsConfig = getConfig().sourceMaps;
  const sourceMapsEnabled = !anymatch(sourceMapsConfig.exclude, filePath) && anymatch(sourceMapsConfig.include, filePath);
  const { debugMode } = getConfig();
  let generatorOptions: GeneratorOptions = {
    sourceMaps: true,
    sourceFileName: fixPath(path.relative(getConfig().rootDir, filePath)),
    sourceRoot: 'reboost:///',
    minified: !debugMode
  }

  const { code: generatedCode, map: generatedMap } = generate(ast, sourceMapsEnabled ? generatorOptions : undefined);
  let map;

  if (sourceMap && sourceMapsEnabled) {
    const merged = await mergeSourceMaps(sourceMap, generatedMap);
    map = getCompatibleSourceMap(merged);
  }

  return {
    code: generatedCode,
    map: map && JSON.stringify(map, null, getConfig().debugMode ? 2 : 0),
    imports,
    dependencies,
    error: errorOccurred
  }
}
