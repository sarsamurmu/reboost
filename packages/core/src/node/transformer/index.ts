import generate, { GeneratorOptions } from '@babel/generator';
import anymatch from 'anymatch';
import { RawSourceMap } from 'source-map';
import chalk from 'chalk';
import MagicString from 'magic-string';

import fs from 'fs';
import path from 'path';

import { PluginContext } from '../index';
import { getConfig, getAddress } from '../shared';
import { mergeSourceMaps, toPosix } from '../utils';
import { resolve } from '../core-plugins/resolver';
import { process } from './processor';
import { resolveImports } from './import-resolver';

const getCompatibleSourceMap = (map: RawSourceMap) => {
  const config = getConfig();

  map.sourceRoot = 'reboost:///';

  map.sources = map.sources.map((sourcePath: string) => {
    if (path.isAbsolute(sourcePath)) return toPosix(path.relative(config.rootDir, sourcePath));
    return toPosix(sourcePath);
  });

  map.sourcesContent = [];
  map.sources.forEach((sourcePath) => {
    const absolutePath = path.join(config.rootDir, sourcePath);
    if (fs.existsSync(absolutePath)) {
      map.sourcesContent.push(fs.readFileSync(absolutePath).toString());
    } else {
      console.log(chalk.red(`Unable to find file "${toPosix(absolutePath)}". Required for source map generation.`));
      map.sourcesContent.push(`Unable to find file in "${toPosix(absolutePath)}".`);
    }
  });

  map.file = undefined;

  return map;
}

const getPluginContext = (filePath: string, mergedDependencies: string[]): PluginContext => ({
  address: getAddress(),
  config: getConfig(),
  addDependency(dependency) {
    mergedDependencies.push(toPosix(dependency));
  },
  chalk,
  getCompatibleSourceMap,
  getSourceMapComment(map) {
    let comment = '/*# sourceMappingURL=data:application/json;charset=utf-8;base64,';
    comment += Buffer.from(JSON.stringify(map)).toString('base64');
    comment += ' */';
    return comment;
  },
  MagicString,
  mergeSourceMaps,
  resolve
});

export const transformFile = async (filePath: string): Promise<{
  code: string;
  dependencies: string[];
  map?: string;
  imports?: string[];
  error?: boolean;
}> => {
  let errorOccurred = false;
  const imports: string[] = [];
  const dependencies: string[] = [];
  const pluginContext = getPluginContext(filePath, dependencies);

  const getErrorObj = (msg: string) => ({
    code: `console.error('[reboost] ' + ${JSON.stringify(msg)})`,
    error: true,
    dependencies
  });

  const processed = await process(filePath, pluginContext);

  if (processed.error) return getErrorObj(processed.error);

  const { ast, sourceMap } = processed;

  errorOccurred = await resolveImports(ast, filePath, imports);

  const sourceMapsConfig = getConfig().sourceMaps;
  const sourceMapsEnabled = !anymatch(sourceMapsConfig.exclude, filePath) && anymatch(sourceMapsConfig.include, filePath);
  const { debugMode } = getConfig();
  const generatorOptions: GeneratorOptions = {
    sourceMaps: true,
    sourceFileName: toPosix(path.relative(getConfig().rootDir, filePath)),
    sourceRoot: 'reboost:///',
    minified: !debugMode
  }

  const { code: generatedCode, map: generatedMap } = generate(ast, sourceMapsEnabled ? generatorOptions : undefined);
  let map;

  if (sourceMap && sourceMapsEnabled) {
    const merged = await mergeSourceMaps(sourceMap, generatedMap);
    map = getCompatibleSourceMap(merged);
  }

  return {
    code: generatedCode,
    map: map && JSON.stringify(map, null, getConfig().debugMode ? 2 : 0),
    imports,
    dependencies,
    error: errorOccurred
  }
}
