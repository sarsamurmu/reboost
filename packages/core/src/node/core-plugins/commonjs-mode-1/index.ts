import { transformCommonJS } from './transform-commonjs';

import { ReboostPlugin } from '../../index';
import { uniqueID } from '../../utils';

export { transformCommonJS as runTransformation }

export const CommonJSMode1Plugin = (): ReboostPlugin => ({
  name: 'core-commonjs-mode-1-plugin',
  getCacheKey: () => 1,
  transformAST(ast, _, filePath) {
    transformCommonJS(ast, filePath, uniqueID(6));
  }
})
