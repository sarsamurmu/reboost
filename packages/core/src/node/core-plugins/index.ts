import { JSONPlugin } from './json';
import { LoaderPlugin } from './loader';
import { ResolverPlugin } from './resolver';
import { CommonJSPlugin } from './commonjs';
import { CommonJSInteropPlugin } from './commonjs-interop';

export const CorePlugins = [
  JSONPlugin,
  LoaderPlugin,
  ResolverPlugin,
  CommonJSPlugin,
  CommonJSInteropPlugin
]
