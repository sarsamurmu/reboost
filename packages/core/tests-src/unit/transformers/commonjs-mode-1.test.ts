import { runTransformation } from 'src-node/core-plugins/commonjs-mode-1';
import { createTransformer } from './transformer';

const t = createTransformer((ast) => runTransformation(ast, 'my-file.js', ''));

describe('transforms CommonJS modules -', () => {
  test('only "exports"', () => {
    expect(t(`
      exports.item1 = 0;
    `)).toMatchSnapshot();

    expect(t(`
      Object.assign(exports, '__esModule', { value: true });
    `)).toMatchSnapshot();
  });

  test('only "module.exports"', () => {
    expect(t(`
      module.exports.item1 = 0;
    `)).toMatchSnapshot('using dot notation');

    expect(t(`
      module['exports'].item1 = 0;
    `)).toMatchSnapshot('using bracket notation');

    expect(t(`
      Object.assign(module.exports, '__esModule', { value: true });
    `)).toMatchSnapshot();
  });

  test('both', () => {
    expect(t(`
      module.exports.item1 = 0;
      exports.item2 = 0;
    `)).toMatchSnapshot();
  });

  test('"require" calls', () => {
    expect(t(`
      const module_1 = require('module_1');
    `)).toMatchSnapshot();
  });
});

describe('does not transform CommonJS modules', () => {
  test('if "module" is defined', () => {
    expect(t(`
      const module = { exports: {} };
      module.exports.item1 = 0;
    `)).toMatchSnapshot();
  });

  test('if "exports" is defined', () => {
    expect(t(`
      const exports = {};
      exports.item1 = 0;
    `)).toMatchSnapshot();
  });

  test('if "require" is defined', () => {
    expect(t(`
      const require = (mod) => someFunc(mod);
      const module_1 = require('module_1');
    `)).toMatchSnapshot();
  });

  test('if "require" function\'s call signature is different', () => {
    expect(t(`
      const module_1 = require('someThing', 0);
      const result = require(100);
    `)).toMatchSnapshot();
  });

  test('if built-in modules are imported using "require"', () => {
    expect(t(`
      const fs = require('fs');
      const path = require('path');
    `)).toMatchSnapshot();
  });
});

describe('transforms ES modules', () => {
  test('which has imports with name', () => {
    expect(t(`
      import Def from 'module_1';
    `)).toMatchSnapshot('default');

    expect(t(`
      import { part1, part2 } from 'module_1';
    `)).toMatchSnapshot('named');

    expect(t(`
      import * as all from 'module_1';
    `)).toMatchSnapshot('all');
  });

  test('which has import just for side effect', () => {
    expect(t(`
      import 'module_1';
    `)).toMatchSnapshot();
  });
});
