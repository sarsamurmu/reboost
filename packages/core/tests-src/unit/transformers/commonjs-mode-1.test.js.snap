// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not transform CommonJS modules if "exports" is defined 1`] = `
"const exports = {};
exports.item1 = 0;"
`;

exports[`does not transform CommonJS modules if "module" is defined 1`] = `
"const module = {
  exports: {}
};
module.exports.item1 = 0;"
`;

exports[`does not transform CommonJS modules if "require" function's call signature is different 1`] = `
"const module_1 = require('someThing', 0);

const result = require(100);"
`;

exports[`does not transform CommonJS modules if "require" is defined 1`] = `
"const require = mod => someFunc(mod);

const module_1 = require('module_1');"
`;

exports[`does not transform CommonJS modules if built-in modules are imported using "require" 1`] = `
"const fs = require('fs');

const path = require('path');"
`;

exports[`transforms CommonJS modules - "require" calls 1`] = `
"import importer_ from \\"#/importer\\";
import * as imported_0_ from \\"module_1\\";
const module_1 = importer_.All(imported_0_, __reboost_resolve(\\"module_1\\"), \\"my-file.js\\");"
`;

exports[`transforms CommonJS modules - both 1`] = `
"const exports = {};
const module = {
  exports
};
module.exports.item1 = 0;
exports.item2 = 0;
export const __cjsExports = module.exports;"
`;

exports[`transforms CommonJS modules - only "exports" 1`] = `
"const exports = {};
exports.item1 = 0;
export const __cjsExports = exports;"
`;

exports[`transforms CommonJS modules - only "module.exports": using bracket notation 1`] = `
"const module = {
  exports: {}
};
module['exports'].item1 = 0;
export const __cjsExports = module.exports;"
`;

exports[`transforms CommonJS modules - only "module.exports": using dot notation 1`] = `
"const module = {
  exports: {}
};
module.exports.item1 = 0;
export const __cjsExports = module.exports;"
`;

exports[`transforms ES modules which has import just for side effect 1`] = `"import 'module_1';"`;

exports[`transforms ES modules which has imports with name: all 1`] = `
"import importer_ from \\"#/importer\\";
import * as import_0_ from \\"module_1\\";
const all = importer_.All(import_0_, __reboost_resolve(\\"module_1\\"), \\"my-file.js\\");"
`;

exports[`transforms ES modules which has imports with name: default 1`] = `
"import importer_ from \\"#/importer\\";
import * as import_0_ from \\"module_1\\";
const Def = importer_.Default(import_0_, __reboost_resolve(\\"module_1\\"), \\"my-file.js\\");"
`;

exports[`transforms ES modules which has imports with name: named 1`] = `
"import importer_ from \\"#/importer\\";
import * as import_0_ from \\"module_1\\";
const part1 = importer_.Member(import_0_, \\"part1\\", __reboost_resolve(\\"module_1\\"), \\"my-file.js\\"),
      part2 = importer_.Member(import_0_, \\"part2\\", __reboost_resolve(\\"module_1\\"), \\"my-file.js\\");"
`;
